src/gafaelfawr/services/token.py

```diff
@@ -0,0 +1,473 @@
+            self._logger.warning("Permission denied", error=msg)
+            raise PermissionDeniedError(msg)
+        await self._token_redis_store.delete(key)
+        with self._transaction_manager.transaction():
+            success = self._token_db_store.delete(key)
+        self._logger.info("Deleted token", key=key)
+        return success
+
+    async def get_data(self, token: Token) -> Optional[TokenData]:
+        """Retrieve the data for a token from Redis.
+
+        Doubles as a way to check the validity of the token.
+
+        Parameters
+        ----------
+        token : `gafaelfawr.models.token.Token`
+            The token.
+
+        Returns
+        -------
+        data : `gafaelfawr.models.token.TokenData` or `None`
+            The data underlying the token, or `None` if the token is not
+            valid.
+        """
+        return await self._token_redis_store.get_data(token)
+
+    def get_info(
+        self, key: str, username: Optional[str] = None
+    ) -> Optional[TokenInfo]:
+        """Get information about a token.
+
+        Parameters
+        ----------
+        key : `str`
+            The key of the token.
+        username : `str`, optional
+            If set, constrain the result to tokens from that user and return
+            `None` if the token exists but is for a different user.
+        """
+        info = self._token_db_store.get_info(key)
+        if not info:
+            return None
+        if username and info.username != username:
+            return None
+        return info
+
+    async def get_internal_token(
+        self, token_data: TokenData, service: str, scopes: List[str]
+    ) -> Token:
+        """Get or create a new internal token.
+
+        The new token will have the same expiration time as the existing token
+        on which it's based unless that expiration time is longer than the
+        expiration time of normal interactive tokens, in which case it will be
+        capped at the interactive token expiration time.
+
+        Parameters
+        ----------
+        token_data : `gafaelfawr.models.token.TokenData`
+            The authentication data on which to base the new token.
+        service : `str`
+            The internal service to which the token is delegated.
+        scopes : List[`str`]
+            The scopes the new token should have.
+
+        Returns
+        -------
+        token : `gafaelfawr.models.token.Token`
+            The newly-created token.
+        """
+        if not set(scopes) <= set(token_data.scopes):
+            raise PermissionDeniedError("Token does not have required scopes")
+
+        # See if there's already a matching internal token.
+        key = self._token_db_store.get_internal_token_key(
+            token_data, service, scopes
+        )
+        if key:
+            data = await self._token_redis_store.get_data_by_key(key)
+            if data:
+                return data.token
```

Comment:
I assume the reissue-if-half-expired check is yet to be implemented.

Reply:
Correct, along with the internal cache.